### PR TITLE
Support launching DropInActivity with a session

### PR DESCRIPTION
**Pull request by jreij**

### CODE REVIEW

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt [hunk 1]
```diff
@@ -173,18 +173,7 @@ class DropInActivity :
...
-                    if (firstPaymentMethod != null) {
-                        showComponentDialog(firstPaymentMethod)
-                    } else {
-                        throw CheckoutException("First payment method is null")
-                    }
-                }
-                dropInViewModel.showPreselectedStored -> showPreselectedDialog()
-                else -> showPaymentMethodsDialog()
-            }
+            dropInViewModel.activityCreated()
```
Comment (OscarSpruit) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt:176]: `onCreated` would be a better name imo. That way it doesn't matter if the view model is used by an activity or fragment and it follows the android naming (e.g. onThis, onThat, onWhateverHappened, etc)

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInActivityEvent.kt [hunk 2]
```diff
@@ -9,11 +9,23 @@
...
 import com.adyen.checkout.components.model.payments.request.OrderRequest
+import com.adyen.checkout.components.model.payments.response.Action
+import com.adyen.checkout.dropin.ui.giftcard.GiftCardPaymentConfirmationData
 
 sealed class DropInActivityEvent {
     data class MakePartialPayment(val paymentComponentState: PaymentComponentState<*>) : DropInActivityEvent()
     object ShowPaymentMethods : DropInActivityEvent()
     class CancelOrder(val order: OrderRequest, val isDropInCancelledByUser: Boolean) : DropInActivityEvent()
     object CancelDropIn : DropInActivityEvent()
+    class LoadFragment(val fragment: DropInFragmentToLoad) : DropInActivityEvent()
```
Comment (OscarSpruit) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInActivityEvent.kt:22]: I would do something like this:
```
...
    class NavigateTo(val destination: DropInDestination) : DropInActivityEvent()
}

internal sealed class DropInDestination {
     object PreselectedStored : DropInDestination()
     object PaymentMethods : DropInDestination()
     class PaymentComponent(val paymentMethod: PaymentMethod) : DropInDestination()
     class ActionComponent(val action: Action) : DropInDestination()
     class GiftCardPaymentConfirmation(val data: GiftCardPaymentConfirmationData) : DropInDestination()
 }
```
That way it's also independent of wether a fragment or activity uses it and it tells a little better what is going to happen imo

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt [hunk 3]
```diff
@@ -112,23 +116,34 @@ class DropInViewModel(
...
+        return paymentMethodsApiResponse?.paymentMethods.orEmpty()
+    }
+
+    fun getStoredPaymentMethods(): List<StoredPaymentMethod> {
+        return paymentMethodsApiResponse?.storedPaymentMethods.orEmpty()
+    }
 
-    val preselectedStoredPayment
-        get() = paymentMethodsApiResponse.storedPaymentMethods?.firstOrNull {
+    fun showPreselectedStored(): Boolean {
```
Comment (OscarSpruit) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt:127]: This name implies that something will be shown, so I would call it `shouldShowPreselectedStored`

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt [hunk 4]
```diff
@@ -137,6 +152,41 @@ class DropInViewModel(
...
+        } else {
+            loadDropIn()
+        }
+    }
+
+    private fun isInitializedWithSession(): Boolean {
+        return initialSession != null
+    }
+
+    fun sessionSetupSuccessful(sessionSetupResponse: SessionSetupResponse) {
```
Comment (OscarSpruit) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt:169]: `onSessionSetupSuccessful `

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt [hunk 5]
```diff
@@ -137,6 +152,41 @@ class DropInViewModel(
...
+        paymentMethodsApiResponse = sessionSetupResponse.paymentMethods
+        loadDropIn()
+    }
+
+    private fun loadDropIn() {
```
Comment (OscarSpruit) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/DropInViewModel.kt:174]: Hmm maybe `showInitialScreen` would be better

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt (hunk 1 above)
Comment (jreij) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt:176]: `onViewCreated` maybe? 🤔

Comment (OscarSpruit) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt]: Hmm that kinda implies fragment, what do you think about `onReady`?

Comment (jreij) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInActivity.kt]: I went with  `onCreated` eventually

**Review by OscarSpruit — APPROVED** [review]

**Pull request merged**